homeassistant/components/sunweg/sensor.py

```diff
@@ -0,0 +1,324 @@
+"""Read status of SunWEG inverters."""
+from __future__ import annotations
+
+import datetime
+from decimal import Decimal
+import json
+import logging
+
+from sunweg.api import APIHelper
+from sunweg.plant import Plant
+
+from homeassistant.components.sensor import SensorEntity
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_NAME, CONF_PASSWORD, CONF_USERNAME
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.typing import StateType
+from homeassistant.util import Throttle
+
+from .const import CONF_PLANT_ID, DEFAULT_PLANT_ID, DOMAIN
+from .sensor_types.inverter import INVERTER_SENSOR_TYPES
+from .sensor_types.phase import PHASE_SENSOR_TYPES
+from .sensor_types.sensor_entity_description import SunWEGSensorEntityDescription
+from .sensor_types.string import STRING_SENSOR_TYPES
+from .sensor_types.total import TOTAL_SENSOR_TYPES
+
+_LOGGER = logging.getLogger(__name__)
+
+SCAN_INTERVAL = datetime.timedelta(minutes=5)
+CACHE_INTERVAL = datetime.timedelta(minutes=1)
+
+
+def get_device_list(api: APIHelper, config):
+    """Retrieve the device list for the selected plant."""
+    plant_id = config[CONF_PLANT_ID]
+
+    # Log in to api and fetch first plant if no plant id is defined.
+    login_response = api.authenticate()
+    if not login_response:
+        _LOGGER.error("Username, Password or URL may be incorrect!")
+        return
+    if plant_id == DEFAULT_PLANT_ID:
+        plant_info: list[Plant] = api.listPlants()
+        plant_id = plant_info[0].id
+
+    # Get a list of devices for specified plant to add sensors for.
+    devices = api.plant(plant_id).inverters
+    return [devices, plant_id]
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the SunWEG sensor."""
+    config = {**config_entry.data}
+    username = config[CONF_USERNAME]
+    password = config[CONF_PASSWORD]
+    name = config[CONF_NAME]
+
+    # Initialise the library with the username & password
+    api = APIHelper(username, password)
```

Comment:
you should access this from coordinator/from hass.data[DOMAIN][config_entry.entry_id]